server/src-lib/Hasura/Cache/Unbounded.hs

```diff
@@ -0,0 +1,90 @@
+{-| An in-memory, unbounded, capability-local cache implementation. By making the cache
+capability-local, data may be recomputed up to once per capability (which usually means up to once
+per OS thread), but write contention from multiple threads is unlikely. -}
+module Hasura.Cache.Unbounded
+  ( UnboundedCache
+  , initialise
+  , clear
+  , insert
+  , lookup
+  , getEntries
+  ) where
+
+import           Hasura.Prelude      hiding (lookup)
+
+import           Control.Concurrent  (getNumCapabilities, myThreadId,
+                                      threadCapability)
+
+import qualified Data.HashMap.Strict as Map
+import qualified Data.IORef          as IORef
+import qualified Data.Vector         as V
+
+newtype LocalCacheRef k v = LocalCacheRef (IORef.IORef (Map.HashMap k v))
+
+getEntriesLocal
+  :: LocalCacheRef k v -> IO [(k, v)]
+getEntriesLocal (LocalCacheRef ioRef) =
+  Map.toList <$> IORef.readIORef ioRef
+
+-- | Create a new LC cache of the given size.
+initialiseLocal :: IO (LocalCacheRef k v)
+initialiseLocal = LocalCacheRef <$> IORef.newIORef Map.empty
+
+clearLocal :: LocalCacheRef k v -> IO ()
+clearLocal (LocalCacheRef ref)=
+  IORef.atomicModifyIORef' ref $ const (Map.empty, ())
+
+lookupLocal :: (Hashable k, Eq k) => LocalCacheRef k v -> k -> IO (Maybe v)
+lookupLocal (LocalCacheRef ref) k =
+  Map.lookup k <$> IORef.readIORef ref
+
+insertLocal :: (Hashable k, Eq k) => LocalCacheRef k v -> k -> v -> IO ()
+insertLocal (LocalCacheRef ref) k v =
+  IORef.atomicModifyIORef' ref $ \c -> (Map.insert k v c, ())
+
+-- | Using a stripe of multiple handles can improve the performance in
+-- the case of concurrent accesses since several handles can be
+-- accessed in parallel.
+newtype UnboundedCache k v = UnboundedCache (V.Vector (LocalCacheRef k v))
+
+getEntries
+  :: UnboundedCache k v -> IO [[(k, v)]]
+getEntries (UnboundedCache localCaches) =
+  mapM getEntriesLocal $ V.toList localCaches
+
+-- | Create a new 'StripedHandle' with the given number of stripes and
+-- the given capacity for each stripe.
+initialise :: IO (UnboundedCache k v)
+initialise = do
+  capabilities <- getNumCapabilities
+  UnboundedCache <$> V.replicateM capabilities initialiseLocal
+
+clear :: UnboundedCache k v -> IO ()
+clear (UnboundedCache caches) =
+  V.mapM_ clearLocal caches
+
+{-# INLINE getLocal #-}
```

Comment:
What is the performance impact of this `INLINE` pragma? It seems unlikely to do much to me, but maybe I’m missing something.